pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion.go

```diff
@@ -41,20 +42,35 @@ import (
 type namespaceDeletionHandler struct {
 	k8sGardenClient kubernetes.Interface
 
-	projectLister gardencorelisters.ProjectLister
-	shootLister   gardencorelisters.ShootLister
-
 	codecs serializer.CodecFactory
 }
 
 // NewValidateNamespaceDeletionHandler creates a new handler for validating namespace deletions.
-func NewValidateNamespaceDeletionHandler(k8sGardenClient kubernetes.Interface, projectLister gardencorelisters.ProjectLister, shootLister gardencorelisters.ShootLister) http.HandlerFunc {
+func NewValidateNamespaceDeletionHandler(ctx context.Context, k8sGardenClient kubernetes.Interface) (http.HandlerFunc, error) {
+	// Initialize caches here to ensure http requests can be served quicker with pre-syncronized caches.
+	var hasSyncFuncs []cache.InformerSynced
+	informer, err := k8sGardenClient.Cache().GetInformerForKind(ctx, corev1.SchemeGroupVersion.WithKind("secret"))
+	if err != nil {
+		return nil, err
+	}
+	hasSyncFuncs = append(hasSyncFuncs, informer.HasSynced)
+	informer, err = k8sGardenClient.Cache().GetInformerForKind(ctx, corev1.SchemeGroupVersion.WithKind("shoot"))
+	if err != nil {
+		return nil, err
+	}
+	hasSyncFuncs = append(hasSyncFuncs, informer.HasSynced)
+
+	for _, hasSync := range hasSyncFuncs {
+		if !hasSync() {
+			return nil, errors.New("timed out waiting for caches to sync")
+		}
+	}
```

Comment:
I think the `HasSynced` funcs themselves don't wait, that's what `cache.WaitForCacheSync` is for, right?
```suggestion
	if !cache.WaitForCacheSync(ctx.Done(), hasSyncFuncs...) {
		return nil, errors.New("timed out waiting for caches to sync")
	}
```

Reply:
Yes, you are right. I pushed it too early with the outstanding comments yesterday. I think we should also combine it with a readiness probe.